### PR TITLE
Remove save-analysis.

### LINKED ISSUE #43606: Tracking issue for `-Zsave-analysis`
-Z flags are now unstable. However, we plan to use save-analysis info for the RLS and Rustdoc and we want these to work on stable compilers. Therefore, we need a stable flag. Options:

* `--save-analysis`
* `--emit analysis` or analysis-info or analysis-data or whatever
* Keep the -Z flag and give tools a loophole to use it on stable

I propose that the existence of the flag is stable, but that we make no guarantees about the output when using the flag. We might consider guaranteeing that the output is valid JSON (but that includes `{}`). We could promise that the shape of the ouput won't change, but I'd rather not.

### PR DESCRIPTION
**Pull request by nnethercote**
Most tests involving save-analysis were removed, but I kept a few where the `-Zsave-analysis` was an add-on to the main thing being tested, rather than the main thing being tested.

Closes https://github.com/rust-lang/rust/issues/43606

### CODE REVIEW

**Comment by nnethercote** [comment]
#43606 is the tracking issue for save analysis removal.

File: src/test/ui/traits/alias/self-in-generics.rs
Comment (bjorn3) [inline src/test/ui/traits/alias/self-in-generics.rs:6]: Please remove extra whitespace.

File: src/test/ui/type-alias-impl-trait/issue-63279.rs
Comment (bjorn3) [inline src/test/ui/type-alias-impl-trait/issue-63279.rs:4]: Same

File: src/test/ui/issues/issue-3763.rs
Comment (lqd) [reply src/test/ui/issues/issue-3763.rs:1]: 69416 was an ICE for save-analysis on a test for issue 3763

File: src/test/ui/feature-gates/feature-gate-associated_type_bounds.rs
Comment (JohnTitor) [inline src/test/ui/feature-gates/feature-gate-associated_type_bounds.rs:2]: This test is for save-analysis and could also be removed.

File: src/test/ui/asm/issue-72570.rs
Comment (JohnTitor) [inline src/test/ui/asm/issue-72570.rs:3]: These are unnecessary too.

File: src/test/ui/methods/assign-to-method.rs
Comment (lqd) [inline src/test/ui/methods/assign-to-method.rs:1]: This test however existed before save-analysis and was repurposed to test a save-analysis ICE at the same time. We can just remove the compile flag and keep the test.

File: src/test/ui/issues/issue-3763.rs
Comment (nnethercote) [reply src/test/ui/issues/issue-3763.rs:1]: Yes. See the filename. This test was added for #3763 originally and then later `-Zsave-analysis` was added in #69416.

File: src/test/ui/feature-gates/feature-gate-associated_type_bounds.rs
Comment (nnethercote) [reply src/test/ui/feature-gates/feature-gate-associated_type_bounds.rs:2]: I tried that and got a tidy error about a missing test for the `associated_type_bound` feature gate.

File: src/test/ui/traits/alias/self-in-generics.rs
Comment (nnethercote) [reply src/test/ui/traits/alias/self-in-generics.rs:6]: I did this in order to avoid changing line numbers in the `.stderr` files. But I can change them if people feel strongly about it.

File: src/test/ui/feature-gates/feature-gate-associated_type_bounds.rs
Comment (JohnTitor) [reply src/test/ui/feature-gates/feature-gate-associated_type_bounds.rs:2]: Ah hm, that makes sense then. Thanks for clarifying!

File: src/test/ui/traits/alias/self-in-generics.rs
Comment (JohnTitor) [reply src/test/ui/traits/alias/self-in-generics.rs:6]: I'd prefer to bless stderr files as it's quite unclear for future contributors why these tests have extra lines.

**Comment by est31** [comment]
Would it be possible to wait with merging of this, maybe 2-3 weeks? Due to this PR, I've switched cargo-udeps over to use binary depinfo, and made the 0.1.33 release, but it needs some time until the update arrives at users. On the other hand cargo-udeps has to be used with nightly rustc, so a removal would get to them very quickly.

**Comment by lqd** [comment]
@est31 there's a bit more work to be done here related to rustup, and will need FCP I think (which lasts 10 at least days or so IIRC), so there's time yeah.

**Comment by nnethercote** [comment]
The perf run shows that bootstrap time was reduced by about 4.6 seconds due to the removal of the `rust_save_analysis` crate.

**Comment by nnethercote** [comment]
I addressed all the comments about tests.

**Comment by nnethercote** [comment]
@rfcbot fcp

**Comment by nnethercote** [comment]
@rfcbot fcp merge

**Comment by eddyb** [comment]
> #43606 is the tracking issue for save analysis removal.

Not super important, but if you want you could add "Closes #43606" to the PR description.

**Comment by nikomatsakis** [comment]
@rfcbot revewied

**Comment by nnethercote** [comment]
> Once a majority of reviewers approve (and at most 2 approvals are outstanding)

With 14 compiler team members, I think this means 12 must approve. We currently have 10 approvals.  (But I just was doing some experimental clicking and it's possible that I may have undone a vote from @estebank, sorry!)

The PR probably can't land in its current form, because something more will be needed for `rustup` to be happy, e.g. leaving behind an empty save-analysis component. I propose waiting until the 12 approvals are in before starting that work, though I'm happy to hear counter-suggestions.

**Comment by nnethercote** [comment]
r? @Mark-Simulacrum 

Most of the changes are straightforward. The bootstrap changes are the trickiest. I removed the `x install` support for `rust-analysis` and changed the `x dist` support so it produces stub output. I think I got it right, but I don't understand that code very well.

**Comment by tmandry** [comment]
Unfortunately Google's [indexing tool](https://github.com/kythe/kythe) depends on save-analysis, and we haven't yet been able to find anyone with time to fix it.

Lately we've investigated using the rust-analyzer tooling for Fuchsia, but all three of rust-analyzer's CLI tools report many panics and then loop infinitely on Fuchsia's codebase, with varying backtraces, in a way that is not easy for us to diagnose or minimize.

Using save-analysis always came with a risk of it being removed, but at the time of developing that tooling, rust-analyzer was not yet in a place where it could be used for this purpose. It seems like there are still some issues we have that need to be chased down. So while I support switching to rust-analyzer (and ultimately removing save-analysis), this does not appear to be as straightforward as I hoped.

In the meantime, if save-analysis is removed we will lose support for the Rust code indexing on https://cs.opensource.google/ and related websites. What's the impact of that? It's not astronomical, but not trivial either. Many developers use the code navigation for getting up to speed in Google's open source code bases that use Rust.

Personally I would much prefer it if we had a longer warning period prior to removal to work out the issues that keep us from adopting rust-analyzer. An announcement e.g. on the Inside Rust blog might also help spread the message to other users who depend on it.

**Comment by wesleywiser** [comment]
Nominating for compiler team discussion. We need to decide when we want to land this change. There's [been an ask on the Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Future.20of.20save-analysis/near/304896312) for us to wait 3 months so that tooling which currently depends on save-analysis can be rewritten to use r-a instead. We should also discuss how much messaging we need to do prior to merging.

**Comment by apiraino** [comment]
Visited during T-compiler meeting, conversation will follow on this [Zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Future.20of.20save-analysis) (also with regard to the question if save-analysis can wait 3 months before being removed, which seems fine).

@rustbot label -I-compiler-nominated

**Comment by nnethercote** [comment]
Three months from today will be February 14, 2023. I will merge this PR on that date or shortly after. That will be one day short of five months after this PR was filed. Thanks to everyone for the discussions so far.

**Comment by nnethercote** [comment]
@Mark-Simulacrum: I have updated. We're getting close the Feb 14 trigger date, should be a good time to review. The tricky part here is the bootstrap stuff and its interaction with rustup, I think.

**Review by Mark-Simulacrum — APPROVED** [review]
Code wise this seems fine to me, I think we can move ahead on Tuesday (?).

I expect it may bounce a few times in CI so feel free to re-approve with minor changes. (r=me)

**Comment by nnethercote** [comment]
I rebased.

@bors r=Mark-Simulacrum

**Comment by nnethercote** [comment]
I rebased again.

@bors r=Mark-Simulacrum

**Pull request merged**
